sdk/patina/src/test.rs

```diff
@@ -307,20 +318,56 @@ mod tests {
         func: |storage| crate::test::__private_api::FunctionTest::new(test_function).run(storage.into()),
     };
 
+    static TEST_CASE3: super::__private_api::TestCase = super::__private_api::TestCase {
+        name: "test_that_fails",
+        skip: false,
+        should_fail: false,
+        fail_msg: None,
+        func: |storage| crate::test::__private_api::FunctionTest::new(test_function_fail).run(storage.into()),
+    };
+
     #[test]
-    fn test_we_run_without_panicking() {
-        if cfg!(feature = "enable_patina_tests") {
-            assert_eq!(2, super::__private_api::test_cases().len());
-        } else {
-            assert_eq!(0, super::__private_api::test_cases().len());
-        }
+    fn test_we_can_initialize_the_component() {
+        let mut storage = Storage::new();
+
+        let mut component = super::TestRunner::default().fail_fast(true).into_component();
+        component.initialize(&mut storage);
+    }
 
+    #[test]
+    fn test_we_can_collect_and_execute_tests() {
+        assert_eq!(TEST_TESTS.len(), 2);
         let mut storage = Storage::new();
+        storage.add_config(1_i32);
 
+        let component = super::TestRunner::default().fail_fast(true);
+        let result = component.run_tests(&TEST_TESTS, &mut storage);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_handle_different_test_counts() {
+        let mut test_cases = vec![];
+        let mut storage = Storage::new();
         storage.add_config(1_i32);
 
-        let mut component = super::TestRunner::default().fail_fast(true).into_component();
-        component.initialize(&mut storage);
-        let _ = component.run(&mut storage);
+        let component = super::TestRunner::default().fail_fast(true);
+        let result = component.run_tests(&test_cases, &mut storage);
+        assert!(result.is_ok());
```

Comment:
general question: if a test fails, will this assert stop execution at this point?

If its possible, I think we would still want to execute the rest of the tests so the developer can get a full list of what failed and be able to fix them in one go

Reply:
The patina-test functionality is for on-system unit tests. It provides `assert!` alternatives - namely `u_assert!`, etc. which instead of panicing, immediately return from the test in error.

On the test runner side, you can configure the test runner (that is the `fail_fast` setting) to either immediately return from the component in failure, or to finish all tests and return in failure at that point.

With that said, changes that will be happening soon remove the `fail_fast` functionality, as we will be allowing users to register unit tests to run on certain events. Since that is going to be added, we are removing the concept of failing fast.